ion_phys/tests/test_tls.py

```diff
@@ -0,0 +1,81 @@
+""" Simple rate equations example of a 2 level system"""
+
+import unittest
+
+from ion_phys.ions.ca43 import Ca43, ground_level, P32
+from ion_phys import Laser
+from ion_phys.rate_equations import Rates
+
+
+def steady_state(intensity):
+    "Steady state poulation in the P-state for resonant intensity /I0"
+    return intensity / (2 * intensity + 1)
+
+
+class TestTLS(unittest.TestCase):
+    """two-level system tests"""
+    def test_steady_state_intensity(self):
+        """Check that the steady state intensity scaling"""
+
+        # use both integers and floats
+        intensity_list = [1e-3, 1e-1, 0.3, 1, 1.0, 2, 10., 1.2e4]
+
+        ion = Ca43(B=5e-4, level_filter=[ground_level, P32])
+        s_idx = ion.index(ground_level, 4)
+        p_idx = ion.index(P32, +5)
+
+        rates = Rates(ion)
+        delta = ion.delta(s_idx, p_idx)
+
+        for I in intensity_list:
+            Lasers = [Laser("393", q=+1, I=I, delta=delta)]  # resonant
+            trans = rates.get_transitions(Lasers)
+
+            Np_ss = steady_state(I)
+            # trqansition rates normalised by A coeficient
+            dNp_dt = (trans[p_idx, p_idx] * Np_ss
+                      + trans[p_idx, s_idx] * (1 - Np_ss))
+            dNp_dt = dNp_dt / (trans[p_idx, p_idx] + trans[p_idx, s_idx])
+            self.assertAlmostEqual(0., dNp_dt, places=7)
+            dNs_dt = (trans[s_idx, p_idx] * Np_ss
+                      + trans[s_idx, s_idx] * (1 - Np_ss))
+            dNs_dt = dNs_dt / (trans[s_idx, p_idx] + trans[s_idx, s_idx])
+            self.assertAlmostEqual(0., dNs_dt, places=7)
+
+    def test_steady_state_detuning(self):
+        """Test steady state detuning dependance"""
+
+        # assume 1 saturation intensity
+        ion = Ca43(B=5e-4, level_filter=[ground_level, P32])
+        s_idx = ion.index(ground_level, 4)
+        p_idx = ion.index(P32, +5)
+
+        rates = Rates(ion)
+        delta = ion.delta(s_idx, p_idx)
+
+        Lasers = [Laser("393", q=+1, I=1., delta=delta)]  # resonant
+        trans = rates.get_transitions(Lasers)
+        line_width = abs(trans[p_idx, p_idx] + trans[p_idx, s_idx])
```

Comment:
Optional: if you wanted to test more of the code base, you could calculate this line-width from 1st principles rather than assuming those elements are correct.